core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -406,410 +365,312 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-      val refsToUpdateAndAccessLevel = refsToUpdate.map { case (ref, (accessLevel, _, _)) => ref -> accessLevel }
-
-      val existingRefsAndSharePermission = existingRefsAndLevels.map { case (ref, (_, canShare, _))  => ref -> canShare }
-      val existingRefsAndComputePermission = existingRefsAndLevels.map { case (ref, (_, _, canCompute))  => ref -> canCompute }
-      val existingRefsAndAccessLevel = existingRefsAndLevels.map { case (ref, (accessLevel, _, _)) => ref -> accessLevel }
-
-      // remove everything that is not changing
-      val actualAccessChangesToMake = refsToUpdateAndAccessLevel.diff(existingRefsAndAccessLevel)
-      val actualShareChangesToMake = refsToUpdateAndSharePermission.filter{ case (_, canShare) => canShare.isDefined }
-        .diff(existingRefsAndSharePermission.map { case (ref, canShare) => ref -> Option(canShare)})
-      val actualComputeChangesToMake = refsToUpdateAndComputePermission.filter{ case (_, canCompute) => canCompute.isDefined }
-        .diff(existingRefsAndComputePermission.map { case (ref, canCompute) => ref -> Option(canCompute)})
-
-      val membersWithTooManyEntries = actualAccessChangesToMake.groupBy {
-        case (member, _) => member
-      }.collect {
-        case (member, entries) if entries.size > 1 => refsToUpdateByEmail.collect {
-          case (email, member2) if member == member2 => email
-        }
-      }.flatten
-
-      if (membersWithTooManyEntries.nonEmpty) {
-        throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"Only 1 entry per email allowed. Emails with more than one entry: $membersWithTooManyEntries"))
-      }
-
-      val membersWithHigherExistingAccessLevelThanGranter = existingRefsAndLevels.filterNot { case (_, (accessLevel, _, _)) => accessLevel == WorkspaceAccessLevels.ProjectOwner }
-        .filter { case (member, _) => actualAccessChangesToMake.map { case(ref, _) => ref }.contains(member) }.filter { case (_, (accessLevel, _, _)) => accessLevel > userAccessLevel }
-
-      if (membersWithHigherExistingAccessLevelThanGranter.nonEmpty) {
-        throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"You may not alter the access level of users with higher access than yourself. Please correct these entries: $membersWithHigherExistingAccessLevelThanGranter"))
-      }
-
-      val membersWithHigherAccessLevelThanGranter = actualAccessChangesToMake.filter { case (_, accessLevel) => accessLevel > userAccessLevel }
-
-      if (membersWithHigherAccessLevelThanGranter.nonEmpty) {
-        throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, s"You may not grant higher access than your own access level. Please correct these entries: $membersWithHigherAccessLevelThanGranter"))
-      }
+  private def validateAclChanges(aclChanges: Set[WorkspaceACLUpdate], existingAcls: Set[WorkspaceACLUpdate]) = {
```

Comment:
I would make this return `Either[RawlsExceptionWithErrorReport, Set[WorkspaceACLUpdate]]` if it's easy refatoring